src/_ert/forward_model_runner/client.py

```diff
@@ -26,107 +22,139 @@ class ClientConnectionClosedOK(Exception):
+        print(f"Created: {self.dealer_id=} {token=} {self._ack_timeout=}")
         if token is not None:
-            self._additional_headers["token"] = token
-
-        # Mimics the behavior of the ssl argument when connection to
-        # websockets. If none is specified it will deduce based on the url,
-        # if True it will enforce TLS, and if you want to use self signed
-        # certificates you need to pass an ssl_context with the certificate
-        # loaded.
-        self._ssl_context: bool | ssl.SSLContext | None = None
-        if cert is not None:
-            self._ssl_context = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
-            self._ssl_context.load_verify_locations(cadata=cert)
-        elif url.startswith("wss"):
-            self._ssl_context = True
-
-        self._max_retries = max_retries
-        self._timeout_multiplier = timeout_multiplier
-        self.websocket: ClientConnection | None = None
+            client_public, client_secret = zmq.curve_keypair()
+            self.socket.curve_secretkey = client_secret
+            self.socket.curve_publickey = client_public
+            self.socket.curve_serverkey = token.encode("utf-8")
+
         self.loop = new_event_loop()
+        self._receiver_task = None
+
+    async def connect(self) -> None:
+        self.socket.connect(self.url)
+        await self._term_receiver_task()
+        self._receiver_task = asyncio.create_task(self._receiver())
+        try:
+            await self._send("CONNECT", retries=1)
+        except ClientConnectionError:
+            await self._term_receiver_task()
+            self.term()
+            raise
+
+    def send(self, message: str, retries: int | None = None) -> None:
+        self.loop.run_until_complete(self._send(message, retries))
+
+    async def process_message(self, msg: str) -> None:
+        pass
+
+    async def _receiver(self) -> None:
+        while True:
+            try:
+                _, raw_msg = await self.socket.recv_multipart()
+                if raw_msg == b"ACK":
+                    self._ack_event.set()
+                else:
+                    await self.process_message(raw_msg.decode("utf-8"))
+            except zmq.ZMQError as exc:
+                logger.debug(
+                    f"{self.dealer_id} connection to evaluator went down, reconnecting: {exc}"
+                )
+                await asyncio.sleep(1)
+                self.socket.connect(self.url)
 
-    async def get_websocket(self) -> ClientConnection:
-        return await connect(
-            self.url,
-            ssl=self._ssl_context,
-            additional_headers=self._additional_headers,
-            open_timeout=self.CONNECTION_TIMEOUT,
-            ping_timeout=self.CONNECTION_TIMEOUT,
-            ping_interval=self.CONNECTION_TIMEOUT,
-            close_timeout=self.CONNECTION_TIMEOUT,
-        )
+    async def _send(self, message: str, retries: int | None = None) -> None:
+        self._ack_event.clear()
```

Comment:
Can we end in a corner case where a forward model step goes from running to finished very quickly, so we try sending a Running and Finished at almost the same time? In this case, will both `send()`-coroutines return when the first "ack" is received?

Reply:
In this case, should every message have its own instance of an ack event (that includes an id) that we store in a dict in the client?

Reply:
In case we will not get `ACK` for a message, then it is guaranteed that no further message is being sent unless we received `ACK` correctly. If the Finished is handled then we have a gracious timeout in Event reporter which tries to send the remaining the messages.